taxcalc/parameters.py

```diff
@@ -1,170 +1,496 @@
-                           " because we don't know how to fill in the "
-                           " values for the years between {0} and {1}.")
-                    raise ValueError(msg.format(self.current_year, year))
-
-            # Set up the '_X = [a, b,...]' variables as 'X = a'
-            self.set_year(self._current_year)
+        if self.current_year != self.start_year:
+            self.set_year(self.start_year)
+        if not reform:
+            return  # no reform to implement
+        reform_years = sorted(list(reform.keys()))
+        last_reform_year = max(reform_years)
+        if last_reform_year > self.end_year:
+            msg = 'reform provision in year={} > end_year={}'
+            ValueError(msg.format(last_reform_year, self.end_year))
+        for year in reform_years:
+            if year != self.start_year:
+                self.set_year(year)
+            self._update({year: reform[year]})
+        self.set_year(self.start_year)
 
     @property
     def current_year(self):
+        """
+        Current policy parameter year property.
+        """
         return self._current_year
 
     @property
     def start_year(self):
+        """
+        First policy parameter year property.
+        """
         return self._start_year
 
     @property
-    def budget_years(self):
-        return self._budget_years
+    def num_years(self):
+        """
+        Number of policy parameter years property.
+        """
+        return self._num_years
 
-    def increment_year(self):
-        self._current_year += 1
-        self.set_year(self._current_year)
+    @property
+    def end_year(self):
+        """
+        Last policy parameter year property.
+        """
+        return self._end_year
 
-    def set_year(self, yr):
-        for name, vals in self._vals.items():
+    def set_year(self, year):
+        """
+        Set policy parameters to values for specified calendar year.
+
+        Parameters
+        ----------
+        year: int
+            calendar year for which to current_year and parameter values
+
+        Raises
+        ------
+        ValueError:
+            if year is not in [start_year, end_year] range.
+
+        Returns
+        -------
+        nothing: void
+
+        Notes
+        -----
+        To increment the current year, use the following statement::
+
+            ppo.set_year(ppo.current_year + 1)
+
+        where ppo is a policy Parameters object.
+        """
+        if year < self.start_year or year > self.end_year:
+            msg = 'year passed to set_year() must be in [{},{}] range.'
+            raise ValueError(msg.format(self.start_year, self.end_year))
+        self._current_year = year
+        year_zero_indexed = year - self._start_year
+        for name in self._vals:
             arr = getattr(self, name)
-            setattr(self, name[1:], arr[yr - self._start_year])
+            setattr(self, name[1:], arr[year_zero_indexed])
+
+    @classmethod
```

Comment:
I would say the main ’pro’ of going with @staticmethod is that it most closely reflects programmer intent for the function. All of the existing use cases for the defaultdata function are essentially about grabbing out the default data that “comes with” taxcalc. It’s just one step above the raw opening of a file and grabbing out JSON data. The ‘con’ of the three new @classmethod functions is that it communicates that there is something stateful about the behavior, or that this behavior is dependent in some way on the Parameters class itself. For example, the style of these three @classmethod functions is desirable if you have a situation where there is a subclass of your base class, and wish to customize behavior of one of your class methods. The advantage of OO kicks in at that point, because the `cls` argument would be the proper class object (either Parameters or its subclass) so you get code reuse. In the existing use cases, we don’t have any need for behavior like that. In the refactor, there are a number of mechanisms in play (several bound methods of the Parameters class, two of which use the convention for “private to this class”) which don’t help communicate the stateless-ness of the function. With a "transplant" of the module-scope function to @staticmethod, you are communicating 'this function is logically grouped with what's happening in this class but it doesn't need the binding mechanism of class methods and I'm just putting it here because if you're reading the code of this class you should probably know what this function does too'. I think that's an accurate picture of the intent of the function.